File: lab_12_langgraph_starter/graph/orchestrator.js
import { StateGraph, Annotation, START, END } from "@langchain/langgraph";
import { createToolCallingAgentNode } from "../agents/tool-calling-agent.js";

/**
 * Shared state for the graph:
 * - userMessage: natural language request (e.g. "Fetch 30 users and give me age and location stats")
 * - finalAnswer: LLM's final response after tool calls
 * - randomUsers: result from fetch_random_users tool (if called)
 * - aggregations: result from compute_statistics tool (if called)
 */
const StateAnnotation = Annotation.Root({
  userMessage: Annotation(),
  finalAnswer: Annotation(),
  randomUsers: Annotation(),
  aggregations: Annotation(),
});

/**
 * Orchestrator: compiles a LangGraph with a single LLM agent that has tools.
 *   START → agent (LLM decides when to call fetch_random_users / compute_statistics) → END
 * The model decides how many users to fetch and which statistics to compute.
 */
export async function createGraph(apiKey) {
  const agentNode = await createToolCallingAgentNode(apiKey);

  const graph = new StateGraph(StateAnnotation)
    .addNode("agent", (state) => agentNode.run(state))
    .addEdge(START, "agent")
    .addEdge("agent", END)
    .compile();

  return {
    graph,
    getInfo: () => ({
      agents: [agentNode.getInfo()],
      flow: "userMessage → agent (LLM + tools: fetch_random_users, compute_statistics) → END",
    }),
  };
}

File: lab_12_langgraph_starter_py/graph/orchestrator.py
"""
Orchestrator: builds a LangGraph with a single LLM agent that has tools.
  START → agent (LLM decides when to call fetch_random_users / compute_statistics) → END
"""
from typing import TypedDict

from langgraph.graph import END, START, StateGraph

from agents.tool_calling_agent import create_tool_calling_agent_node


class State(TypedDict, total=False):
    user_message: str
    final_answer: str | None
    random_users: list | None
    aggregations: list | None


def create_graph(api_key: str) -> dict:
    agent_node = create_tool_calling_agent_node(api_key)

    def agent_node_fn(state: State) -> dict:
        return agent_node["run"](state)

    builder = StateGraph(State)
    builder.add_node("agent", agent_node_fn)
    builder.add_edge(START, "agent")
    builder.add_edge("agent", END)
    graph = builder.compile()

    def get_info() -> dict:
        return {
            "agents": [agent_node["get_info"]()],
            "flow": "user_message → agent (LLM + tools: fetch_random_users, compute_statistics) → END",
        }

    return {"graph": graph, "get_info": get_info}
